system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the JS file backend and JS file of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "wasmfs.h"
+
+using js_index_t = uint32_t;
+
+extern "C" {
+void _emscripten_write_js_file(js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset);
+void _emscripten_read_js_file(js_index_t index,
+                              const uint8_t* buffer,
+                              size_t length,
+                              off_t offset);
+int _emscripten_get_js_file_size(js_index_t index);
+int _emscripten_create_js_file();
+void _emscripten_remove_js_file(js_index_t index);
+}
+
+namespace wasmfs {
+
+// This class describes a file that lives in JS Memory
+class JSFile : public DataFile {
+  // This index indicates the location of the JS File in the backing JS array.
+  js_index_t index;
+
+  // JSFiles will write from a Wasm Memory buffer into the backing JS array.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+
+    _emscripten_write_js_file(index, buf, len, offset);
+
+    return __WASI_ERRNO_SUCCESS;
```

Comment:
We should eventually have error handling here. For example, failing to allocate a new typed array will throw in JS, and we should catch that and return a write error I think. But for now, just a TODO seems ok.

Reply:
Is there a good way to propagate errors from JS to C++? @kripken . I added the TODO in the meantime.

Reply:
I suppose we could do the same thing that system calls do, which is to return the error as a negative number.

Reply:
Done!